src/cluster_import.c

```diff
@@ -0,0 +1,1927 @@
+}
+
+void clusterHandleSlotImportLinkClientClose(slotMigrationLink *link) {
+    link->client = NULL;
+    if (!isSlotMigrationLinkInProgress(link)) return;
+    if (link->state == SLOT_IMPORT_RECONNECT) return;
+    clusterDoBeforeSleep(CLUSTER_TODO_HANDLE_SLOT_MIGRATION);
+    serverLog(LL_WARNING,
+              "Slot import link %.40s lost connection to node %.40s "
+              "(owner of slots [%s]). Reconnecting...",
+              link->linkname,
+              link->nodename,
+              link->slot_ranges_str);
+    updateSlotMigrationLinkState(link, SLOT_IMPORT_RECONNECT);
+}
+
+void clusterHandleSlotMigrationLinkClientOOM(void *o) {
+    slotMigrationLink *link = (slotMigrationLink *)o;
+    if (link->type != SLOT_MIGRATION_IMPORT) return;
+    if (isSlotMigrationLinkInProgress(link)) {
+        serverLog(LL_WARNING,
+                  "Slot import link %.40s to node %.40s (owner of slots [%s]) "
+                  "failed due due to OOM",
+                  link->linkname,
+                  link->nodename,
+                  link->slot_ranges_str);
+        finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED,
+                                "Ran out of memory (OOM) during slot import");
+    }
+}
+
+/* Validate that the slot migration is still owned by the node it originally
+ * was importing from. If the source node has since changed, or no longer has
+ * sole ownership by the original source node, we fail the migration. Otherwise,
+ * return the source node for this migration. */
+clusterNode *validateClusterNodeForImportIsUnchanged(slotMigrationLink *link) {
+    serverAssert(link->type == SLOT_MIGRATION_IMPORT);
+    int cross_node;
+    clusterNode *n = getClusterNodeBySlotRanges(link->slot_ranges, &cross_node);
+    if (n && !memcmp(n->name, link->nodename, CLUSTER_NAMELEN)) {
+        return n;
+    }
+    if (n == server.cluster->myself) {
+        serverLog(LL_WARNING,
+                  "Slot import link %.40s to node %.40s (previous owner of slots [%s]) "
+                  "already owned by myself. Failing slot import...",
+                  link->linkname,
+                  link->nodename,
+                  link->slot_ranges_str);
+        finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED,
+                                "Slots were unexpectedly assigned to myself during import");
+        return NULL;
+    }
+    serverLog(LL_WARNING,
+              "Slot import link %.40s to node %.40s (previous owner of slots [%s]) "
+              "contains at least one slot no longer owned by the source node. "
+              "Failing slot import...",
+              link->linkname,
+              link->nodename,
+              link->slot_ranges_str);
+    finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED,
+                            "Slots are no longer owned by source node");
+    return NULL;
+}
+
+void performSlotImportLinkReconnect(slotMigrationLink *link) {
```

Comment:
More of a personal preference, I prefer the way that the replication handles it where there is no "reconnecting" state. Cluster import links are put into connecting, and when they fail they are cleaned up in place and either put back into connecting or deleted.